crates/hotshot/src/traits/networking/push_cdn_network.rs

```diff
@@ -0,0 +1,424 @@
+                Marshal::new(marshal_config)
+                    .await
+                    .expect("failed to spawn marshal");
+
+            // Error if we stopped unexpectedly
+            if let Err(err) = marshal.start().await {
+                error!("broker stopped: {err}");
+            }
+        });
+
+        // This function is called for each client we spawn
+        Box::new({
+            move |node_id| {
+                // Derive our public and priate keys from our index
+                let private_key =
+                    TYPES::SignatureKey::generated_from_seed_indexed([0u8; 32], node_id).1;
+                let public_key = TYPES::SignatureKey::from_private(&private_key);
+
+                // Calculate if we're DA or not
+                let topics = if node_id < da_committee_size as u64 {
+                    vec![Topic::DA, Topic::Global]
+                } else {
+                    vec![Topic::Global]
+                };
+
+                // Configure our client
+                let client_config = ClientConfigBuilder::default()
+                    .keypair(KeyPair {
+                        public_key: WrappedSignatureKey(public_key),
+                        private_key,
+                    })
+                    .subscribed_topics(topics)
+                    .endpoint(marshal_endpoint.clone())
+                    .build()
+                    .expect("failed to build client config");
+
+                // Create our client
+                let client = Arc::new(PushCdnNetwork(
+                    async_block_on(async move { Client::new(client_config).await })
+                        .expect("failed to create client"),
+                ));
+
+                (client.clone(), client)
+            }
+        })
+    }
+
+    /// Get the number of messages in-flight.
+    ///
+    /// Some implementations will not be able to tell how many messages there are in-flight. These implementations should return `None`.
+    fn in_flight_message_count(&self) -> Option<usize> {
+        None
+    }
+}
+
+#[async_trait]
+impl<TYPES: NodeType> ConnectedNetwork<Message<TYPES>, TYPES::SignatureKey>
+    for PushCdnNetwork<TYPES>
+{
+    /// We do not support pausing the PushCDN network right now, but it is possible.
+    fn pause(&self) {}
+
+    /// We do not support resuming the PushCDN network right now, but it is possible.
+    fn resume(&self) {}
+
+    /// The clients form an initial connection when created, so we don't have to wait.
+    async fn wait_for_ready(&self) {}
+
+    /// The clients form an initial connection when created, so we can return `true` here
+    /// always.
+    async fn is_ready(&self) -> bool {
+        true
+    }
+
+    /// TODO: shut down the networks. Unneeded for testing.
+    fn shut_down<'a, 'b>(&'a self) -> BoxSyncFuture<'b, ()>
+    where
+        'a: 'b,
+        Self: 'b,
+    {
+        // TODO: figure out if drop is sufficient
+        boxed_sync(async move {})
```

Comment:
I didn't look into the implementation of the client too much, but I assume it's spawning tasks that won't get shutdown properly unless we implement here.  Is that the case, or is it actually possible that dropping the client is enough.  

I'm mostly worried because in CI all the tasks we don't shutdown will continue for the whole test run and it could pollute other tests and cause some flakiness.  It might also be fine.  If it's difficult to implement no reason to block the PR.  Mostly just curious what will happen now during shutdown.

Reply:
The drop inside the client should be sufficient. I will do some testing tomorrow, but we have `Drop` implemented for the underlying connection so that when a reference there gets dropped, the underlying channels get terminated and the underlying task gets stopped as a result.